R/utils.R

```diff
@@ -140,3 +140,107 @@ edit.data.table = function(name, ...) {
   setDT(NextMethod('edit', name))[]
 }
 # nocov end
+
+col_helper = function(x, colsub, origin_call = NULL) {
+  ## helper that could replace some common code used for by, .SDcols, j, duplicated, unique, dcast, 
+  ## or anything that looks for columns
+  
+  #' @param x required; a data.frame, data.table, tibble, or list.
+  #' @param colsub takes the same arguments as .SDcols such as
+  #' `patterns()`, `var1:var2`, a variable, or a vector of characters.
+  #' When TRUE, will return seq_len(length(x))
+  #' @param origin_call takes character vector of length 1. In the short-term
+  #' there will be different modes based on the call. In the long-term, it's just
+  #' to help the end-user what caused the error
+  #' @return Integer vector with an attribute of whether negation was in the call
+  
+  # see @jangorecki #4174 for how to get the vast majority of this into C
+  
+  x_len = length(x)
+  
+  if (length(colsub) == 1L && isTRUE(colsub)){
+    ansvals = seq_len(x_len)
+    attr(ansvals, "negate") = FALSE
+    return(ansvals)
+  }
+  
+  # FR #4979 - negative numeric and character indices for SDcols
+  # fix for #5190. colsub[[1L]] gave error when it's a symbol.
+  negate_cols = is.call(colsub) && deparse(colsub[[1L]], 500L, backtick=FALSE) %in% c("!", "-")
+  if (negate_cols) colsub = colsub[[2L]]
+  
+  if (is.call(colsub)){
+    # fix for #1216, make sure the parentheses are peeled from expr of the form (((1:4)))
+    while(colsub %iscall% "(") colsub = as.list(colsub)[[-1L]]
+    if (length(colsub) == 3L && colsub[[1L]] == ":") {
```

Comment:
Maybe switch to `%chin%`? Since Matt added support to avoid coercion of `SYMSXP` to `STRSXP` IIUC. Worth benchmarking maybe.

Reply:
Keeping as is - ```colsub[[1L]] == ":"``` is an order of magnitude faster. Still, I don't see benchmarks using nanoseconds that frequently.
```r
e = quote(1:3)

Unit: nanoseconds
                                expr  min   lq mean median   uq   max neval
                      e[[1L]] == ":"  400  500  647    600  700  2600   100
 data.table:::"%chin%"(e[[1L]], ":") 5200 5500 6126   5700 5900 39500   100

Reply:
Based on other conversation, my benchmark is likely faulty. Still, I doubt ```%chin%``` would improve performance over the current 600 nanoseconds. I am re-resolving changing ```colsub[[1L]] == ":"``` to ```colsub[[1L]] %chin% ":"```